File: java/src/main/java/com/sasteval/vuln/hard/CommandRunner.java
package com.sasteval.vuln.hard;

import java.io.IOException;

/**
 * CWE-78: OS Command Injection (Cross-File Helper - Command Runner)
 * Part of a 3-file taint chain: CommandInjectionCrossFile -> CommandBuilder -> CommandRunner.
 */
public class CommandRunner {

    public Process runCommand(String cmd) throws IOException {
        // VULN: Executes shell command string without sanitization
        return Runtime.getRuntime().exec(new String[]{"/bin/sh", "-c", cmd});
    }
}

File: dotnet/Vuln/Hard/CommandRunner.cs
using System.Diagnostics;

namespace SastEval.Vuln.Hard;

/// <summary>
/// Helper class for CommandInjectionCrossFile - executes shell command.
/// </summary>
public static class CommandRunner
{
    public static string Run(string cmd)
    {
        // VULNERABLE: Executes arbitrary shell command built from user input
        var process = Process.Start(new ProcessStartInfo
        {
            FileName = "/bin/sh",
            Arguments = "-c " + cmd,
            RedirectStandardOutput = true,
            UseShellExecute = false
        });

        return process?.StandardOutput.ReadToEnd() ?? "No output";
    }
}
